File: internal/domain/task.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type TaskStatus string

const (
	TaskStatusNew        TaskStatus = "new"
	TaskStatusInProgress TaskStatus = "in_progress"
	TaskStatusDone       TaskStatus = "done"
)

func (s TaskStatus) IsValid() bool {
	switch s {
	case TaskStatusNew, TaskStatusInProgress, TaskStatusDone:
		return true
	}
	return false
}

type Task struct {
	ID          uuid.UUID  `db:"id"`
	Title       string     `db:"title"`
	Description string     `db:"description"`
	Status      TaskStatus `db:"status"`
	Deadline    *time.Time `db:"deadline"`
	CreatedAt   time.Time  `db:"created_at"`
	UpdatedAt   time.Time  `db:"updated_at"`
	DeletedAt   *time.Time `db:"deleted_at"`
	AuthorID    uuid.UUID  `db:"author_id"`
}

func (t *Task) IsOverdue() bool {
	if t.Deadline == nil || t.Status == TaskStatusDone {
		return false
	}
	return time.Now().After(*t.Deadline)
}

// TaskFilter — параметры фильтрации и пагинации списка задач.
type TaskFilter struct {
	Status         *TaskStatus
	Search         *string
	Overdue        *bool
	DeadlineBefore *time.Time
	DeadlineAfter  *time.Time
	CreatedAfter   *time.Time
	CreatedBefore  *time.Time
	SortBy         string // created_at, deadline, status, title
	Order          string // asc, desc
	Page           int
	Limit          int
}

// TaskStats — агрегированная статистика по задачам пользователя.
type TaskStats struct {
	Total    int            `json:"total"`
	ByStatus map[string]int `json:"by_status"`
	Overdue  int            `json:"overdue"`
}

type TaskRepository interface {
	Create(ctx context.Context, task *Task) error
	GetByID(ctx context.Context, id uuid.UUID) (*Task, error)
	Update(ctx context.Context, task *Task) error
	SoftDelete(ctx context.Context, id uuid.UUID) error
	List(ctx context.Context, authorID uuid.UUID, filter *TaskFilter) ([]Task, int, error)
	GetStats(ctx context.Context, authorID uuid.UUID) (*TaskStats, error)
}

File: web/src/types/task.ts
export type TaskStatus = 'new' | 'in_progress' | 'done'

export interface Task {
  id: string
  title: string
  description: string
  status: TaskStatus
  deadline?: string
  created_at: string
  updated_at: string
  author_id: string
}

export interface CreateTaskRequest {
  title: string
  description?: string
  deadline?: string
}

export interface UpdateTaskRequest {
  title?: string
  description?: string
  status?: TaskStatus
  deadline?: string
}

export interface TaskStats {
  total: number
  by_status: Record<string, number>
  overdue: number
}

export interface TaskFilter {
  status?: TaskStatus
  search?: string
  overdue?: boolean
  deadline_before?: string
  deadline_after?: string
  sort_by?: 'created_at' | 'deadline' | 'status' | 'title'
  order?: 'asc' | 'desc'
  page?: number
  limit?: number
}
